Refactor fracture domain integrals to use vectorpostprocessors

bwspenc opened the pull request:
closes #9559 

The system for fracture integrals was developed before VectorPostprocessors (VPPs) were introduced in MOOSE. These integrals involve computations of quantities along lines on the the crack front, so the VPP system is a much more natural way to do this than the current way, which creates a (large!) set of Postprocessors.

bwspenc committed:
Convert fracture integrals to VectorPostprocessors ref #9559

bwspenc committed:
Add VectorPostprocessorComponent Postprocessor ref #11439

This is a Postprocessor that extracts a component of a
VectorPostprocessor. Adds code, tests, and documentation.

bwspenc committed:
Get FEType from displacement variables rather than assuming FIRST/LAGRANGE ref #9559

Also resolve a concern that tangential strain for T-stress might not be
parallel consistent.

bwspenc committed:
Document fracture integrals ref #9559

bwspenc committed:
Add Doxygen to fracture integral classes ref #9559

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/454888/) on c16a1e1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14621/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14621/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d1fbae50f134b3771146d5647c373d011b77eaa1`

bwspenc committed:
Clang-format fracture integral code ref #9559

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/455155/) on 04a64f3 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14621/site)

This comment will be updated on new commits.

bwspenc committed:
More doxygen and documentation of fracture integrals ref #9559

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/456927/) on ad16a40 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14621/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14621/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d79b9185bcd10b1a9e19c3b0b01ca46bd5794c52`

bwspenc committed:
Fix compile issues on some platforms ref #9559

dschwen commented on modules/solid_mechanics/include/vectorpostprocessors/InteractionIntegralSM.h:
```diff
@@ -76,6 +76,14 @@ class InteractionIntegralSM : public ElementIntegralPostprocessor
+    Distance
+  };
+
+  const PositionType _position_type;
```
those two statements could be combined into one

dschwen commented on modules/solid_mechanics/src/vectorpostprocessors/InteractionIntegralSM.C:
```diff
@@ -123,42 +132,33 @@ InteractionIntegralSM::initialSetup()
+  RealVectorValue crack_direction(1.0, 0.0, 0.0);
 
   // Calculate (r,theta) position of qp relative to crack front
   Point p(_q_point[_qp]);
```
Why is a copy of `_q_point[_qp]` created here?

dschwen commented on modules/solid_mechanics/src/vectorpostprocessors/InteractionIntegralSM.C:
```diff
@@ -263,27 +259,81 @@ InteractionIntegralSM::computeIntegral()
   unsigned int ring_base = (_q_function_type == QMethod::Topology) ? 0 : 1;
 
-  for (unsigned int i = 0; i < _current_elem->n_nodes(); ++i)
+  for (unsigned int icfp = 0; icfp < _interaction_integral.size(); icfp++)
```
I thought we were moving away form `unsigned int` loop indices. At the very least I'd use `std::size_t`

dschwen commented on modules/solid_mechanics/src/vectorpostprocessors/InteractionIntegralSM.C:
```diff
@@ -263,27 +259,81 @@ InteractionIntegralSM::computeIntegral()
-  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
-    sum += _JxW[_qp] * _coord[_qp] * computeQpIntegral();
-  return sum;
+  for (auto i = beginIndex(_interaction_integral); i < _interaction_integral.size(); ++i)
```
or, like here, beginIndex

bwspenc commented on modules/solid_mechanics/src/vectorpostprocessors/InteractionIntegralSM.C:
```diff
@@ -106,8 +114,9 @@ InteractionIntegralSM::InteractionIntegralSM(const InputParameters & parameters)
-    mooseError(
-        "The number of variables supplied in 'displacements' must match the mesh dimension.");
+    mooseError("InteractionIntegralSM Error: number of variables supplied in 'displacements' must "
+               "match the mesh dimension.");
```
The problem with paramError is that it puts a little string on the end (can't remember what it is) that doesn't really make sense after this sentence.

bwspenc replied:
I just tested this.  I'm not crazy.  It tacks this 'not present in InputParams' to the end of the string in line 877 of InputParameters.h:
```
*** ERROR ***
param 'DomainIntegral error: must set radius_inner and radius_outer.' not present in InputParams

bwspenc replied:
Looks like if you use the version of `paramError()` that takes two arguments, you get a better looking message. That version only makes sense when it's related to only one parameter, though.

dschwen replied:
It doesn't tack a message on! And there is no single argument version of `paramError`. The function is always interpreting the first argument as the parameter name - which it takes to be `DomainIntegral error: must set radius_inner and radius_outer.` in your example - which would be a ridiculous name for a parameter, so yeah... it obviously is not present in InputParams :-)

The case I pointed out is really only related to a single parameter: `displacements` (which I actually gave an example for in my comment above...

Your point about combinations of multiple parameters throwing an error is taken though. I've encountered that before and always tried to pick one of the parameters to tag with the error.

bwspenc replied:
I just pushed an update that converts the instances of mooseError that make sense to paramError.

bwspenc committed:
Address review comments ref #9559

bwspenc commented:
@dschwen I just pushed an update that I believe addresses all of the issues you pointed out in the files that I touched. All except converting `mooseError()` to `paramError()`, that is. That one requires some more investigation.

moosebuild commented:
Job [App tests](https://civet.inl.gov/job/458567/) on b91bea4 : invalidated by @permcody

rattlesnake/Mammoth updated

moosebuild commented:
Job [App tests PETSc alt](https://civet.inl.gov/job/458568/) on b91bea4 : invalidated by @permcody

rattlesnake/Mammoth updated

bwspenc committed:
Convert mooseError to paramError ref #9559
